File: .claude/skills/agent-skill-kit-commons/packages/py/src/skills_ref/audit/scripts.py
from __future__ import annotations

import json
import re
from pathlib import Path

_PATTERNS_PATH = Path(__file__).resolve().parent.parent / "assets" / "injection_patterns.json"

_PATTERNS = None


def _patterns():
    global _PATTERNS
    if _PATTERNS is None:
        raw = json.loads(_PATTERNS_PATH.read_text(encoding="utf-8"))
        _PATTERNS = [
            {
                "id": p["id"],
                "re": re.compile(p["pattern"]),
                "languages": set(p["languages"]),
            }
            for p in raw
        ]
    return _PATTERNS


def _walk_scripts(scripts_dir: Path) -> list:
    out = []
    try:
        entries = sorted(scripts_dir.iterdir(), key=lambda p: p.name)
    except OSError:
        return out
    for entry in entries:
        if entry.is_dir():
            out.extend(_walk_scripts(entry))
        elif entry.is_file():
            out.append(entry)
    return out


def audit_scripts(registry) -> list:
    pats = _patterns()
    findings = []
    for name, node in registry.nodes.items():
        d = node.get("dir")
        if not d:
            continue
        scripts_dir = Path(d) / "scripts"
        if not scripts_dir.is_dir():
            continue
        for file in _walk_scripts(scripts_dir):
            ext = file.suffix
            applicable = [p for p in pats if ext in p["languages"]]
            if not applicable:
                continue
            try:
                text = file.read_text(encoding="utf-8")
            except OSError:
                continue
            lines = text.split("\n")
            for i, line in enumerate(lines):
                for p in applicable:
                    if p["re"].search(line):
                        findings.append(
                            {
                                "severity": "HIGH",
                                "code": "A010_COMMAND_INJECTION_PATTERN",
                                "where": f"{file}:{i + 1}",
                                "message": (
                                    f"command-injection pattern '{p['id']}' matched: "
                                    f"{line.strip()}"
                                ),
                            }
                        )
    return findings

File: .claude/skills/agent-skill-kit-commons/packages/mjs/src/audit/scripts.mjs
import { readFileSync, readdirSync, statSync } from "node:fs";
import { join, extname, dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));
const PATTERNS_PATH = resolve(__dirname, "..", "..", "assets", "injection_patterns.json");

let _patterns = null;
function loadPatterns() {
  if (_patterns) return _patterns;
  const raw = JSON.parse(readFileSync(PATTERNS_PATH, "utf8"));
  _patterns = raw.map((p) => ({
    id: p.id,
    re: new RegExp(p.pattern),
    languages: new Set(p.languages),
  }));
  return _patterns;
}

function walkScripts(scriptsDir) {
  const out = [];
  let entries;
  try {
    entries = readdirSync(scriptsDir, { withFileTypes: true });
  } catch {
    return out;
  }
  entries.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));
  for (const entry of entries) {
    const p = join(scriptsDir, entry.name);
    if (entry.isDirectory()) {
      out.push(...walkScripts(p));
    } else if (entry.isFile()) {
      out.push(p);
    }
  }
  return out;
}

/**
 * Scan every file under each skill's scripts/ for canonical command-injection
 * patterns. Each match emits HIGH A010_COMMAND_INJECTION_PATTERN with
 * `path:line` (1-indexed) provenance.
 */
export function auditScripts(registry) {
  const patterns = loadPatterns();
  const findings = [];
  for (const [, node] of registry.nodes) {
    if (!node.dir) continue;
    const scriptsDir = join(node.dir, "scripts");
    let stat;
    try {
      stat = statSync(scriptsDir);
    } catch {
      continue;
    }
    if (!stat.isDirectory()) continue;
    for (const file of walkScripts(scriptsDir)) {
      const ext = extname(file);
      const applicable = patterns.filter((p) => p.languages.has(ext));
      if (applicable.length === 0) continue;
      let text;
      try {
        text = readFileSync(file, "utf8");
      } catch {
        continue;
      }
      const lines = text.split("\n");
      for (let i = 0; i < lines.length; i++) {
        const line = lines[i];
        for (const p of applicable) {
          if (p.re.test(line)) {
            findings.push({
              severity: "HIGH",
              code: "A010_COMMAND_INJECTION_PATTERN",
              where: `${file}:${i + 1}`,
              message: `command-injection pattern '${p.id}' matched: ${line.trim()}`,
            });
          }
        }
      }
    }
  }
  return findings;
}
